cookbooks/cdo-varnish/test/integration/cloudfront/minitest/test_default.rb

```diff
@@ -0,0 +1,315 @@
+
+    # Origin sees request cookie
+    response = request url, {}, {cookie_key: 'cookie_value'}
+    assert_equal text_cookie, last_line(response)
+
+    # Origin sets response cookie
+    response = request url
+    assert_equal text, last_line(response)
+    refute_nil /Cookie: [^\s]+/.match(response)
+
+    # Cache strips request cookie and strips response cookie
+    response = proxy_request url, {}, {cookie_key: 'cookie_value'}
+    assert_equal text, last_line(response)
+    assert_nil /Cookie: [^\s]+/.match(response)
+
+    # Cache hit on changed cookies
+    assert_hit proxy_request url, {}, {cookie_key: 'cookie_value3', key2: 'value2'}
+  end
+
+  it 'Allows whitelisted request cookie to affect cache behavior' do
+    url = get_url 5
+    cookie = "_learn_session_#{ENVIRONMENT}"
+    text = 'Hello World!'
+    text_cookie = 'Hello Cookie 123!'
+    text_cookie_2 = 'Hello Cookie 456!'
+    mock_url(url, text)
+    mock_url(url, text_cookie, {'Cookie' => "#{cookie}=123;"})
+    mock_url(url, text_cookie_2, {'Cookie' => "#{cookie}=456;"})
+
+    # Cache passes request cookie to origin
+    response = proxy_request url, {}, {"#{cookie}" => '123'}
+    assert_equal text_cookie, last_line(response)
+    assert_miss response
+
+    # Cache miss on changed cookies
+    response = proxy_request url, {}, {"#{cookie}" => '456'}
+    assert_equal text_cookie_2, last_line(response)
+    assert_miss response
+  end
+
+  it 'Strips non-whitelisted request cookies' do
+    url = get_url 6
+    cookie = 'random_cookie'
+    text = 'Hello World!'
+    text_cookie = 'Hello Cookie!'
+    mock_url(url, text, {})
+    mock_url(url, text_cookie, {'Cookie' => "#{cookie}=123;"}, {'Set-Cookie' => "#{cookie}=456; path=/"})
+
+    # Request without cookie
+    response = proxy_request url
+    assert_equal text, last_line(response)
+    assert_miss response
+
+    # Cache strips non-whitelisted request cookie and hits original cached response
+    response = proxy_request url, {}, {"#{cookie}" => '123'}
+    assert_equal text, last_line(response)
+    assert_hit response
+  end
+
+  it 'Strips non-whitelisted response cookies' do
+    # TODO: not implemented in CloudFront
+    skip 'Not implemented in CloudFront'
+
+    url = get_url 7
+    cookie = 'random_cookie'
+    text = 'Hello World!'
+    mock_url(url, text, {}, {'Set-Cookie' => "#{cookie}=abc; path=/"})
+
+    # Response should have cookie stripped
+    response = proxy_request url
+    assert_equal text, last_line(response)
+    assert_nil /Set-Cookie: [^\s]+/.match(response)
+    assert_miss response
+
+    # Response should be cached even if response cookie is changed
+    mock_url(url, text, {}, {'Set-Cookie' => "#{cookie}=def; path=/"})
+    assert_hit proxy_request url
+  end
+
+  it 'Does not strip cookies from uncached PUT or POST asset requests' do
+    # Varnish only; not supported by CloudFront
+    skip 'Not implemented in CloudFront'
```

Comment:
It's not possible for this test to succeed through CloudFront + Varnish, only through Varnish alone. Since CloudFront does not separate its cookie-filtering based on the HTTP request method, it filters non-whitelisted cookies from PUT/POST requests as well as from GET/HEAD requests.